### PR TITLE
Reputation

### PR DESCRIPTION
**Pull request by Vid201**
Several sanity checks and simulation conditions are based on the reputation of entities (paymasters, senders, aggregators, factories). Thus we should have a reputation implementation first.

This PR contains:

- reputation management
- debug endpoints
- small changes to the Makefile

### CODE REVIEW

File: bin/bundler.rs [hunk 1]
```diff
@@ -87,6 +90,13 @@ fn main() -> Result<()> {
                             ))
                             .await?;
 
+                            #[cfg(debug_assertions)]
```
Comment (zsluedem) [inline bin/bundler.rs:93]: Is this flag necessary? I think we would still need this in production build.

File: src/uopool/mod.rs [hunk 2]
```diff
@@ -18,18 +25,20 @@ use jsonrpsee::tracing::info;
...
-pub mod memory;
+pub mod memory_mempool;
+pub mod memory_reputation;
 pub mod server;
 pub mod services;
 
 pub type MempoolId = H256;
 
+pub type MempoolBox<T> = Box<dyn Mempool<UserOperations = T>>;
+pub type ReputationBox<T> = Box<dyn Reputation<ReputationEntries = T>>;
```
Comment (zsluedem) [inline src/uopool/mod.rs:36]: I am curious why this needs to be `dyn` now. Why not just use `MemoryMempool` and `MemoryReputation`.

**Review by zsluedem — COMMENTED** [review]
Good work! I left some comments and questions.

File: bin/bundler-rpc.rs [hunk 3]
```diff
@@ -34,6 +37,14 @@ async fn main() -> Result<()> {
     let mut api = Methods::new();
     let uopool_grpc_client =
         UoPoolClient::connect(format!("http://{}", opt.uopool_grpc_listen_address)).await?;
+
+    #[cfg(debug_assertions)]
```
[the remark this answers is not in the document]
Comment (Vid201) [reply bin/bundler-rpc.rs:41]: Debug API should be available only for passing the bundler spec tests. In production release, they must be turned off.

File: src/uopool/mod.rs [hunk 4]
```diff
@@ -53,24 +102,57 @@ pub struct UserOperationPool<M: Mempool> {
...
             mempools.insert(id, Box::<MemoryMempool>::default());
+
+            reputations.insert(id, Box::<MemoryReputation>::default());
+            if let Some(reputation) = reputations.get_mut(&id) {
+                reputation.init(
+                    MIN_INCLUSION_RATE_DENOMINATOR,
+                    THROTTLING_SLACK,
+                    BAN_SLACK,
+                    opts.min_stake,
+                    opts.min_unstake_delay,
```
[the remark this answers is not in the document]
Comment (Vid201) [reply src/uopool/mod.rs:135]: Yes, the first three are config parameters that are the same between all different entry points.

File: src/uopool/mod.rs (hunk 2 above)
Comment (Vid201) [reply src/uopool/mod.rs:36]: Well, my idea was that sometimes, someone would want to use `MemoryMempool` for one mempool and `DatabaseMempool` for another. But at the moment, this is somewhat complex, yeah.

Comment (zsluedem) [reply src/uopool/mod.rs:36]: Hm. Ok. Let's leave it as it is now.

File: src/uopool/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (zsluedem) [reply src/uopool/mod.rs:135]: Cools

**Review by zsluedem — APPROVED** [review]
Nice work.

**Pull request merged**